### PR TITLE
removed tpm2_abrmd, added env variable, and created docker file

### PR DESCRIPTION
**Pull request by bu3alwa**
solves #27 and updates readme

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -50,6 +50,32 @@ For TPM 1.2 use:
...
+
+docker build -t tpm-emulator-base -f docker/Dockerfile .
+```
+
+To run a docker container from the built image
+```
+
+docker run -it -d --name=keylime1 \
+            -p 8443:8443 \
+            -v /path/to/keylime:/keylime \
```
Comment (lukehinds) [inline README.md:69]: Might be better to attach to /root/keylime..e.g.

```
docker run -it -d --name=keylime1 \
            -p 8443:8443 \
            -v /path/to/keylime:/root/keylime \
            tpm-emulator-base
```

**Comment by bu3alwa** [comment]
@lukehinds thanks, it seems `tpm_server` is running and it stops `tpm_serverd` from executing. I'll look into why it's running.

**Comment by bu3alwa** [comment]
So I looked into it and `tpm_serverd` is just a script that runs `tpm_server` 
```
#!/bin/bash

/usr/bin/pgrep -x "tpm_server" &> /dev/null
if [[ $? -eq 0 ]]; then
   echo "tpm_server is already running, stop it first with pkill -x tpm_server"
   exit 1
fi

export TPM_PATH=~/.tpm0/

mkdir -p ${TPM_PATH}

touch ${TPM_PATH}/tpm_server.log

echo "$(date) TPM Server Starting" >> ${TPM_PATH}/tpm_server.log

tpm_server -rm >> ${TPM_PATH}/tpm_server.log &

sleep 2
```
On the vagrant box, the same behavior happens and only `tpm_server` is visible and not `tpm_serverd`

Also, I can find tpm0 does this mean that its running fine?

image

**Comment by lukehinds** [comment]
If it works better, you could extract the operations from the `tpm_serverd` script. This might be a good idea, as we don't need to be concerned over checking if a tpm_server is already running. We used `tpm_serverd` as a development tool originally for when needing a log and clear state. 

So currently we have two forms of TPM. The emulator which runs  as a sort of network service and secondly a real hardware TPM, which is a device that presents itself as `/dev/tpm0`. This is why we set the `TCTI_TOOLS` environment variable. This tells the tools (`tpm2_tools`) how to connect to the TPM. For this we can use `tpm2_abrmd` which is a resource manager and helps with session management towards the TPM, or a kernel based TPM resource manager, which we are starting g to favour as it means one less code base upstream that could break on us. 

So we have `tpm2_abrmd`:

export TPM2TOOLS_TCTI="tabrmd:bus_name=com.intel.tss2.Tabrmd"

This is turn can be run to connect to the emulator:

`/usr/sbin/tpm2-abrmd --tcti=mssim`

* note: mssim is from when Microsoft originally wrote the emulator and later open sourced it. 

Or we can connect directly to the emulator:

`export TPM2TOOLS_TCTI="mssim:port=2321"`

or directly to the HW TPM

`export TPM2TOOLS_TCTI="device:/dev/tpm0"`

These are the steps you would use to check the emulator is healthy:

Set your TCTI 

`export TPM2TOOLS_TCTI="mssim:host=localhost,port=2321"`

Run the server:

`tpm_serverd` or `tpm_server -rm &` 

Set the TPM into the start up state:

`tpm2_startup -c`

Run some commands, for example get 8 random bits from the TPM:

`tpm2_getrandom 8`

Or you can look at the platform configuration registrars:

`tpm2_pcrread`

These are a mix of sha1 and sha256 hashes. These are where the tpm measures a system component and lodges the hash of that component into a secure store called a PCR / platform configuration register.

You can set these within `tpm_policy` in `/etc/keylime.conf` or push it to a Agent.

For example, if you looked at PCR8 this is the grub / kernel command line options in the current state that you would like it to remain in (keep in mind someone can turn off SELinux by changing the grub command line that is stored before booting a system). By setting this hash into PCR8's value, you can remotely have your machine rebooted and once it comes on again Keylime will check that someone has not messed with your box and loaded a trojan or turned off security controls.

Likewise for PCR10 we measure the whole file system into here using something in the Linux Kernel called IMA (Integrity Measurement Architecture). Each time someone then executes a binary, script or changes the contents of a file, the TPM will measure that file (in much the same way as running `sha256sum anyfile.txt` and store it within the TPM. Keylime than requests a measurement of PCR10 every second or so and using the public part of an RSA cryptographic key that is stored only within the TPM, it can make sure no one has tampered with the file system.

You can see this happening here on the demo vid I put together:

https://www.youtube.com/watch?v=Qhr_aVBCZPw&feature=youtu.be&t=586

There is a lot to get your head around with TPMs and how they work, so ask away until it makes sense. We have a 'nothing is a stupid question' policy on the project, so never be concerned to ask when something does not make sense.

**Comment by lukehinds** [comment]
Oh I forgot to mention about `sys/class/tpm*`

The files in `/dev` are actual devices files which UDEV creates at run time, this is where you would find a hardware TPM, but not a software based emulator. The directory `/sys/class` is exported by the kernel at run time, exposing the hierarchy of the hardware through `sysfs`. For the TPM use, this means you can grep out certain things from sysfs such as the TPM version.

**Comment by lukehinds** [comment]
Platform Configuration Registers:

!()[https://i.imgur.com/SzFK9U0.png]

**Comment by lukehinds** [comment]
Last of all, a bit more about PCR10 and IMA:

https://keylime.dev/blog/2019/04/02/running-IMA-on-keylime.html

**Comment by bu3alwa** [comment]
ok so if we do 
```
docker run -it -d --name=keylime1 \
            -p 8443:8443 \
            -v /path/to/keylime:/root/keylime \
            tpm-emulator-base /usr/local/bin/tpm_server -rm
```

the docker container will run with the emulator and then if we need to drop to bash we just `docker  exec -it /bin/bash`

Do you think this is a good solution for the docker container?

**Pull request closed without merging**